File: src/data/ImageEncodeData.cpp
//
// Created by Andrei Costinescu (andreicostinescu96@gmail.com) on 29-May-21.
//

#include <comm/data/ImageEncodeData.h>
#include <cassert>
#include <utility>

using namespace comm;
using namespace std;

const int ImageEncodeData::headerSize = 6 * sizeof(int);

string ImageEncodeData::convertEncodingToString(Encoding encoding) {
    switch (encoding) {
        case JPEG: {
            return "jpg";
        }
        case PNG: {
            return "png";
        }
        case TIFF: {
            return "tiff";
        }
        default : {
            throw runtime_error("Unknown encoding " + to_string(encoding));
        }
    }
}

ImageEncodeData::Encoding ImageEncodeData::convertStringToEncoding(const string &encoding) {
    if (encoding == "jpg") {
        return Encoding::JPEG;
    } else if (encoding == "png") {
        return Encoding::PNG;
    } else if (encoding == "tiff") {
        return Encoding::TIFF;
    } else {
        throw runtime_error("Unknown encoding " + encoding);
    }
}

ImageEncodeData::ImageEncodeData() : ImageData(), encoding(Encoding::PNG), encodedContentSize(0), encodedImage() {}

#ifdef WITH_OPENCV
ImageEncodeData::ImageEncodeData(cv::Mat image, int id, Encoding encoding) :
        ImageData(std::move(image), id), encoding(encoding), encodedContentSize(0), encodedImage() {}
#endif

ImageEncodeData::ImageEncodeData(const vector<uchar> &imageEncodedBytes, int imageHeight, int imageWidth, int imageType,
                                 int id, Encoding encoding) :
        ImageData(nullptr, 0, imageHeight, imageWidth, imageType, id), encoding(encoding),
        encodedContentSize(0), encodedImage(imageEncodedBytes) {
    this->encodedContentSize = this->encodedImage.size();
}

MessageType ImageEncodeData::getMessageType() {
    return MessageType::IMAGE_ENCODE;
}

bool ImageEncodeData::serialize(Buffer *buffer, int start, bool forceCopy, bool verbose) {
    switch (this->serializeState) {
        case 0: {
            buffer->setBufferContentSize(start + ImageEncodeData::headerSize);
            buffer->setInt(this->id, start);
            buffer->setInt(this->imageHeight, start + 4);
            buffer->setInt(this->imageWidth, start + 8);
            buffer->setInt(this->imageType, start + 12);

            #ifdef WITH_OPENCV
            if (this->imageBytes != nullptr) {
                cv::imencode("." + ImageEncodeData::convertEncodingToString(this->encoding), this->image,
                             this->encodedImage);
                this->encodedContentSize = this->encodedImage.size();
            }
            #endif

            buffer->setInt(this->encodedContentSize, start + 16);
            buffer->setInt(this->encoding, start + 20);
            if (verbose) {
                cout << "Serialize: " << this->imageHeight << ", " << this->imageWidth << ", " << this->imageType
                     << ", " << this->encodedContentSize << endl;
                char *dataBuffer = buffer->getBuffer();
                cout << "Serialized content: ";
                for (int i = 0; i < ImageEncodeData::headerSize; i++) {
                    cout << ((int) (unsigned char) dataBuffer[start + i]) << " ";
                }
                cout << endl;
            }
            this->serializeState = 1;
            return false;
        }
        case 1: {
            buffer->setBufferContentSize(this->encodedContentSize);
            char *writeBuffer = buffer->getBuffer();
            for (int i = 0; i < this->encodedContentSize; i++) {
                writeBuffer[i] = this->encodedImage[i];
            }
            this->serializeState = 0;
            return true;
        }
        default: {
            (*cerror) << "Impossible serialize state... " << this->serializeState << endl;
            this->resetSerializeState();
            return false;
        }
    }
}

int ImageEncodeData::getExpectedDataSize() const {
    switch (this->deserializeState) {
        case 0: {
            return ImageEncodeData::headerSize;
        }
        case 1: {
            return this->encodedContentSize;
        }
        default : {
            throw runtime_error("Impossible deserialize state... " + to_string(this->deserializeState));
        }
    }
}

char *ImageEncodeData::getDeserializeBuffer() {
    return nullptr;
}

bool ImageEncodeData::deserialize(Buffer *buffer, int start, bool forceCopy, bool verbose) {
    switch (this->deserializeState) {
        case 0: {
            this->imageDeserialized = false;
            this->id = buffer->getInt(start);
            this->imageHeight = buffer->getInt(start + 4);
            this->imageWidth = buffer->getInt(start + 8);
            this->imageType = buffer->getInt(start + 12);
            this->encodedContentSize = buffer->getInt(start + 16);
            this->encoding = Encoding(buffer->getInt(start + 20));
            this->deserializeState = 1;
            return false;
        }
        case 1: {
            this->encodedImage.resize(this->encodedContentSize);
            assert((buffer->getBufferContentSize() - start) == this->encodedContentSize);

            char *encodedBuffer = buffer->getBuffer();
            for (int i = 0; i < this->encodedContentSize; i++) {
                this->encodedImage[i] = encodedBuffer[i];
            }
            #ifdef WITH_OPENCV
            // TODO: change the flag depending on the image type!
            cv::imdecode(this->encodedImage, cv::IMREAD_COLOR, &this->image);
            if (this->image.data != nullptr) {
                this->imageDeserialized = true;
            }
            #else
            this->imageDeserialized = true;
            #endif

            this->deserializeState = 0;
            return true;
        }
        default: {
            (*cerror) << "Impossible deserialize state... " << this->deserializeState << endl;
            this->resetDeserializeState();
            return false;
        }
    }
}

void ImageEncodeData::setEncoding(Encoding _encoding) {
    this->encoding = _encoding;
}

ImageEncodeData::Encoding ImageEncodeData::getEncoding() {
    return this->encoding;
}

void ImageEncodeData::setImageEncodedBytes(const std::vector<uchar> &imageEncodedBytes, int imageHeight, int imageWidth,
                                           int imageType, int id, Encoding _encoding) {
    this->id = id;
    this->setImage(nullptr, 0, imageHeight, imageWidth, imageType);
    this->encodedImage = imageEncodedBytes;
    this->encodedContentSize = this->encodedImage.size();
    this->encoding = _encoding;
}

File: python/comm/comm_data/ImageEncodeData.py
import cv2 as cv
import enum
import numpy as np
from comm.comm_data.ImageData import ImageData
from comm.comm_data.MessageType import MessageType
from comm.comm_utils.Buffer import Buffer
from comm.comm_utils.utils import memcpy
from typing import Optional


class ImageEncodeData(ImageData):
    headerSize = 24

    class Encoding(enum.Enum):
        JPEG = 0,
        PNG = 1,
        TIFF = 2

        @staticmethod
        def convertEncodingToInt(encoding: 'Encoding'):
            if encoding == ImageEncodeData.Encoding.JPEG:
                return 0
            elif encoding == ImageEncodeData.Encoding.PNG:
                return 1
            elif encoding == ImageEncodeData.Encoding.TIFF:
                return 2
            else:
                raise RuntimeError("Unknown encoding " + str(encoding))

        @staticmethod
        def convertIntToEncoding(encoding: int):
            if encoding == 0:
                return ImageEncodeData.Encoding.JPEG
            elif encoding == 1:
                return ImageEncodeData.Encoding.PNG
            elif encoding == 2:
                return ImageEncodeData.Encoding.TIFF
            else:
                raise RuntimeError("Unknown encoding " + str(encoding))

        @staticmethod
        def convertEncodingToString(encoding: 'Encoding'):
            if encoding == ImageEncodeData.Encoding.JPEG:
                return "jpg"
            elif encoding == ImageEncodeData.Encoding.PNG:
                return "png"
            elif encoding == ImageEncodeData.Encoding.TIFF:
                return "tiff"
            else:
                raise RuntimeError("Unknown encoding " + str(encoding))

        @staticmethod
        def convertStringToEncoding(encoding: str):
            if s == "jpeg":
                return ImageEncodeData.Encoding.JPEG
            elif s == "png":
                return ImageEncodeData.Encoding.PNG
            elif encoding == "tiff":
                return ImageEncodeData.Encoding.TIFF
            else:
                raise RuntimeError("Unknown Encoding " + encoding)

    def __init__(self, image: np.ndarray = None, id: int = -1, encoding: Encoding = None):
        super(ImageEncodeData, self).__init__(image, id)
        self.encodedImage = []  # type: np.ndarray
        self.encodedContentSize = 0
        self.encoding = encoding

    def getMessageType(self) -> MessageType:
        return MessageType.IMAGE_ENCODE

    def serialize(self, buffer: Buffer, start: int, forceCopy: bool, verbose: bool) -> bool:
        if self.serializeState == 0:
            buffer.setBufferContentSize(ImageData.headerSize)
            if verbose:
                print("Serialize: ", self.image.shape[0], ", ", self.image.shape[1], ", ", self.image.shape[2], ", ",
                      self.image.size)

            buffer.setInt(self.id, start)
            buffer.setInt(self.imageHeight, start + 4)
            buffer.setInt(self.imageWidth, start + 8)
            buffer.setInt(self.imageType, start + 12)

            retval, self.encodedImage = cv.imencode(
                "." + ImageEncodeData.Encoding.convertEncodingToString(self.encoding), self.image);
            self.encodedContentSize = len(self.encodedImage)

            buffer.setInt(self.encodedContentSize, start + 16)
            buffer.setInt(ImageEncodeData.Encoding.convertEncodingToInt(self.encoding), start + 20)

            if verbose:
                dataBuffer = buffer.getBuffer()
                print("Serialized content: ")
                for i in range(ImageData.headerSize):
                    print(int(dataBuffer[i]), "", end="")
                print()
            self.serializeState = 1
            return False
        elif self.serializeState == 1:
            buffer.setBufferContentSize(self.encodedContentSize)
            buffer.buffer = memcpy(buffer.getBuffer(), start, self.encodedImage.tobytes(), 0, self.encodedContentSize)
            self.serializeState = 0
            return True
        else:
            print("Impossible serialize state...", self.serializeState)
            self.resetSerializeState()
            return False

    def getExpectedDataSize(self) -> int:
        if self.deserializeState == 0:
            return ImageEncodeData.headerSize
        elif self.deserializeState == 1:
            return self.encodedContentSize
        else:
            raise RuntimeError("Impossible deserialize state... " + str(self.deserializeState))

    def getDeserializeBuffer(self) -> Optional[bytes]:
        return None

    def deserialize(self, buffer: Buffer, start: int, forceCopy: bool, verbose: bool) -> bool:
        if self.deserializeState == 0:
            self.imageDeserialized = False
            self.id = buffer.getInt(start)
            self.imageHeight = buffer.getInt(start + 4)
            self.imageWidth = buffer.getInt(start + 8)
            self.imageType = buffer.getInt(start + 12)
            self.encodedContentSize = buffer.getInt(start + 16)
            self.encoding = ImageEncodeData.Encoding.convertIntToEncoding(buffer.getInt(start + 20))

            # print("Deserializing image:", self.imageHeight, self.imageWidth, self.imageType, self.contentSize)
            self.deserializeState = 1
            return False
        elif self.deserializeState == 1:
            self.encodedImage = np.frombuffer(buffer.getBuffer(), dtype=np.int8)
            self.image = cv.imdecode(self.encodedImage, cv.IMREAD_COLOR)
            self.imageHeight = self.image.shape[0]
            self.imageWidth = self.image.shape[1]
            self.imageType = ImageData.imageCVType(self.image)
            self.imageDeserialized = True
            self.deserializeState = 0
            return True
        else:
            print("Impossible deserialize state...", self.deserializeState)
            self.resetSerializeState()
            return False

    def setEncoding(self, _encoding: Encoding):
        self.encoding = _encoding

    def getEncoding(self):
        return self.encoding
